### PR TITLE
Add LuaSec SSL/TLS support

### PR DESCRIPTION
**Pull request by bcandersen**
Adding [LuaSec](https://github.com/brunoos/luasec/) SSL/TLS support to cosock

I wasn't exactly sure how to fit it in as the ssl library is not part of LuaSocket. Perhaps there should be a `cossl` directory containing `ssl.lua` and a separate `cossl.lua`? I thought that was a bit heavy for this but can certainly restructure however desired.

I added all of the LuaSec API's I could find in the source as there is no API reference doc. Could certainly be missing a few.

Added two test files that mimic the tcp tests. The tests obviously need test keys and certs so I added a shell script that uses openssl to generate everything necessary. A note was added in the README on how gen the keys and certs. If desired, I can just check in the generated files instead. Certs are good for 10 years.

Signed-off-by: Barry Andersen <barry.andersen@smartthings.com>

### CODE REVIEW

File: cosock/cosocket/internals.lua [hunk 1]
```diff
@@ -10,7 +10,8 @@ function m.passthroughbuilder(recvmethods, sendmethods)
         local ret = {isock[method](isock, ...)}
         local status = ret[1]
         local err = ret[2]
-        if err == "timeout" then
+        -- "wantread" is returned by ssl on timeout of dohandshake or receive
+        if err == "timeout" or err == "wantread" then
```
Comment (azdle) [inline cosock/cosocket/internals.lua:14]: Should this also check for a `wantwrite` if a send method would block?

Comment (azdle) [reply cosock/cosocket/internals.lua:14]: Hmm, it sounds like dohandshake can block as both a reader or a writer: https://github.com/brunoos/luasec/wiki/LuaSec-0.9#conn_dohandshake

If I'm reading that right we might need to change the logic the determines `kind` and which parameter to yield into. I would personally change `recvmethods` and `sendmethods` from `method = true` to `method = "err"`, then check that here. That changes our assumption from a method can only have one `kind` to a method always only has one error per `kind`. Which I think is still true.

**Review by azdle — COMMENTED** [review]
Also, the tests should get added to https://github.com/cosock/cosock/blob/master/.travis.yml

**Comment by azdle** [comment]
Oh and for the file naming stuff. I think I'd put it in a file called `ssl.lua` in the `cosock` folder. So that `require "cosock.ssl"` works. The `cosocket` folder really needs to get renamed to just `socket` so that `require "cosock.socket"` also works. I plan to do that before the next release.

File: cosock/cosocket/internals.lua (hunk 1 above)
Comment (bcandersen) [reply cosock/cosocket/internals.lua]: I did not realize the existence of that API reference... :facepalm:. You're right `wantread` and `wantsend` are used for `dohandshake` and they also appear to be used for `send` and `receive` as well. I think I got pretty close to what you are looking for with passing in the expected method error.

**Comment by bcandersen** [comment]
> Also, the tests should get added to https://github.com/cosock/cosock/blob/master/.travis.yml

Added tests to travis.yml and checked in required keys and certs needed for tests.

File: cosock/cosocket/internals.lua (hunk 1 above)
Comment (azdle) [reply cosock/cosocket/internals.lua]: Exactly what I had in mind. :+1:

**Review by azdle — APPROVED** [review]

**Comment by azdle** [comment]
Tested manually, not going to wait for travis to run our CI.

**Pull request merged**